README.md

```diff
@@ -43,39 +50,46 @@ This is essentially a time-to-live (TTL) for the differential update. It dictate
 
 We propose using the [RCS format](https://www.gnu.org/software/diffutils/manual/diffutils.html#RCS) for the diff files. This format is widely used in the software development industry and is well documented. It is also supported by the `patch` utility which is available on most operating systems.
 
-In order to support batch updates and validation of the result, the format is extended with the `f` directive:
-`f [filename] [sha1sum] [numlines]`
+In order to support batch updates and be able to validate patch result, the standard format is extended with the `diff` directive:
 
-* `filename` - a name or a relative path of the file to be patched. See batch updates below for more information on this.
-* `sha1sum` - the expected sha1sum of the file after the patch is applied. This is used to validate the patch.
-* `numlines` - the number of lines that follow that make up the RCS diff block. Note, that `numlines` are counted using the same algorithm as used by `wc -l`, i.e. it basically counts `\n`.
+`diff name:[name] checksum:[checksum] lines:[lines]`
 
-**IMPORTANT** The `f` directive is optional.
+* `name` - name of a corresponding filter list (see `Diff-Name`).
+* `checksum` - the expected SHA1 checksum of the file after the patch is applied. This is used to validate the patch.
+* `lines` - the number of lines that follow that make up the RCS diff block. Note, that `lines` are counted using the same algorithm as used by `wc -l`, i.e. it basically counts `\n`.
+
+`diff` directive is optional. If it is not specified, the patch is applied without validation.
+
+> It is recommended to use the `diff checksum:` directive to validate the patching result. This will ensure that the patch is applied correctly and the resulting file is not corrupted.
 
 ### Algorithm
 
-#### Check for Update
+#### 1. Check for Update
 
 1. Refer to the `Diff-Path` to see if a differential update is available.
+    * If there are several lists with the same `Diff-Path`, download the diff file only once. Refer to the [Batch Updates](#batch-updates) section for the details on how batch patches are applied.
 1. If the differential update is available, download and apply it to the current filter list.
     * Once the differential update is applied, the `Diff-Path` within the list will be updated to point to the next differential update.
     * At this point the ad blocker may decide either to wait for the `Diff-Expires` period and then try again or to immediately try to fetch the next differential update.
 1. If the differential update is not available, i.e. the server returns `404 Not Found`, then the ad blocker wait for the `Diff-Expires` period and then try again.
```

Comment:
I am not a server person, so I wouldn't know, but is it possible returning en empty file is cheaper than a 404? If so, we should allow to return an empty file to be equivalent of "patch does not exist".

Reply:
You know what, after giving it a thought, 404 is doable but I'd prefer to have a 200 OK alternative. Maybe just an empty 200 OK file.

The reasoning for that is that it's hard to predict how different CDNs handle 404. Some CDNs may serve a custom 404 page and it may be rather heavy. Some CDNs like ours may even return 403: https://filters.adtidy.org/somerandomfile which is not cached according to response files.

Empty file is easier to manage: it is cached (but with almost every CDN you can purge cache whenever you need) and it is smaller than any 404 page.

Reply:
> you can purge cache whenever you need

Forgot about this, we don't really control that aspect for the CDNs used in uBO, so the way it's going, the empty file would be cached for much longer I would want for a differential update to occur.

> smaller than any 404 page

Probably a `HEAD` request would work? Negative side is that if the patch exists, server has to deal with two requests instead of one.

Reply:
> Forgot about this, we don't really control that aspect for the CDNs used in uBO, so the way it's going, the empty file would be cached for much longer I would want for a differential update to occur.

Hm, are you talking about jsdelivr? It's there actually: https://www.jsdelivr.com/tools/purge
Maybe @jimaek has time to chime in and add contribute some real expertise here? :)

The point about caching duration of an empty file vs 404, there's really no clear winner here as it all depends on the CDN. For example, Cloudflare caches 404 responses for the same duration as 200. Github pages (Fastly probably?) cache TTL is 10 minutes and I don't think they have different TTL for 404 and 200.

Among the CDNs that you're using the only problem is `statically.io` and tbh I'd argue that you shouldn't be using it for serving lists with differential updates (or maybe even not use it at all). First of all, they use a huge cache control for CDN resources (from 1 day to 1 year) which is a big problem by itself. On top of that they provide no option for purging cache anymore: https://statically.io/docs/purging-cache/. Basically, you'll always have serious inconsistencies between the list versions served from `statically.io` and other CDNs.

Anyways, I am not suggesting replacing 404 in the spec, just adding an alternative option. Basically, instead of counting only 404 as a valid response we could have several types there:

* 200 OK with empty content
* 204 No Content
* 404 Not Found

> Probably a HEAD request would work? Negative side is that if the patch exists, server has to deal with two requests instead of one.

Yeah, the problem with `HEAD` is that it's an extra request. Also, it does not help with inconsistencies in how different CDNs handle 404, some may cache it, some won't, etc.

Reply:
404s may indeed show different caching behavior than other content. In the case of jsDelivr, they are only cached for a very short time (usually 1 minute) so that we're able to pick up new files quickly. However, that also means you shouldn't request too many things that result in a 404 on purpose/too often because the responses will be slower (often uncached) and result in a higher load for us - it's fine if it's going to be only a few unique URLs, though.

Reply:
Also, note that in case of a file that existed at some point and then started to return a 404, we may decide to return the older version instead of the error, in some cases. This currently applies only when requesting tagged releases, not commits/branches. Other providers may have similar features - the idea here is that a 404 for something that worked usually means something broke and we want to minimize the impact, but in your case, that might be undesirable.